gradslide/zosa.py

```diff
@@ -0,0 +1,335 @@
+        # Lim = 24
+        # if u_norm > Lim:
+        #     u_next *= Lim/u_norm
+
+        u_tilde = (1.0-theta_t) * u_tilde + theta_t * u_next
+
+        u = u_next
+
+        u_est = (1.0-g_k)*(x_post) + g_k*(u_tilde)
+        print('in_gap={:.3e}'.format(f_eval(u_est) - fstar))
+        print('in_consensus={:.3e}'.format(cons(u_est)/R))
+
+    return [u, u_tilde]
+
+if len(sys.argv) < 7:
+    print('python zosa.py --graph {1,2,3} --prob {1,2,3,4} --start {1,2,3} [--simple]')
+    print('\t{1,2,3}={chain,random,complete}, {1,2,3}={Geomed, SVM, Lin+l1, Log+l1}, {1,2,3}={random,fixed,zero}')
+    exit(0)
+
+seed_num = 0
+if len(sys.argv)>=9 and sys.argv[7]=='--seed':
+    seed_num = int(sys.argv[8])
+
+[graph_mode,prob_mode, x_0_mode] = [int(i) for i in sys.argv[2:7:2]]
+simple_l1 = False
+
+######## SETUP #################################
+np.random.seed(seed_num)
+n = 100
+
+# Geometric median
+if prob_mode == 1:
+    n = 10
+    m = 100
+    b = np.array([np.random.normal(loc=0, scale=10.0, size=n) for i in range(m)])
+    b = np.reshape(b, newshape=(-1,))
+
+    def df(x): return df_gm_long(x,b,m)
+    def f_eval(x): return f_gm(x,b,m)
+
+    xstar = gm_opt(b,m)
+    fstar = f_gm_comb(xstar,b,m)
+    xstar = np.kron(xstar, np.ones(m))
+
+    M = 1
+
+# SVM 
+elif prob_mode == 2:
+    b, X = readin_csv('wdbc.data')
+    X = X.T
+    reg = 'l1'
+
+    d,m = X.shape
+    n = d
+
+    def df(theta): return df_svm(theta, X, b, reg)
+    def f_eval(theta): return f_svm_long(theta, X, b, reg)
+
+    c = 0.1
+    M = c*((n*m)**0.5)
+
+    xstar = svm_opt(X, b, reg)
+    fstar = f_svm(xstar, X, b, reg)
+
+# Linear Reg with l1
+elif prob_mode == 3:
+    reg = 'l1'
+    m = n
+    d = 10
+    n = d
+
+    mean = 10*np.random.random()-5
+    var  = 10.0
+    X = np.array([np.random.normal(loc=mean, scale=var, size=d) for _ in range(m)]).T
+    y = np.dot(X.T, np.ones(d)) + np.cos(np.dot(X.T, np.ones(d))) + np.random.normal(loc=0, scale=0.25, size=m)
+    u = np.ones(d)
+    y = np.dot(X.T,u)
+
+    def df_smooth(theta): return df_regls(theta, X, y, reg=None)
+    # def df(theta): return np.sign(theta)
+    def df(theta): return df_regls(theta, X, y, reg=reg)
+    def f_eval(theta): return f_regls_long(theta, X, y, reg=reg)
+
+    c = 0.1
+    M = c*((n*m)**0.5)
+    # eigenvalue approach
+    eig_max = np.amax(la.eig(np.dot(X,X.T))[0].real)
+    L = eig_max/m
+
+    # each element approach
+    # X_norms = la.norm(X, ord=2, axis=0)**2
+    # L = (2/m)*(np.amax(X_norms)+c)
+
+    xstar = regls_opt(X, y, reg=reg)
+    fstar = f_regls(xstar, X, y, reg=reg)
+    xstar = np.kron(np.ones(m), xstar)
+
+    # simple_l1 = True
+
+# Logistic Reg with l1
+elif prob_mode == 4:
+    reg = 'l1'
+    m = n
+    d = 10
+    n = d
+    y = np.append(2*np.ones(m//2), np.zeros(m-m//2))-1
+    X = np.array([np.random.normal(loc=y[i]*np.ones(d), scale=1.0, size=d) for i in range(m)]).T
+
+    # TODO: Why can't we solve inner loop 
```

Comment:
Is this to be addressed?

Reply:
Short answer, no. Prox-slide solves an optimization problem in the inner-most loop, and I was testing whether it can be solved exactly when the non-smooth term is simple, e.g., is l_1 regularization. It didn't work when I tested it.